resolver-dns/src/main/java/io/netty/resolver/dns/BiDnsQueryLifecycleObserver.java

```diff
@@ -0,0 +1,96 @@
+package io.netty.resolver.dns;
+
+import io.netty.channel.ChannelFuture;
+import io.netty.handler.codec.dns.DnsQuestion;
+import io.netty.handler.codec.dns.DnsResponseCode;
+import io.netty.util.internal.UnstableApi;
+
+import java.net.InetSocketAddress;
+import java.util.List;
+
+import static io.netty.util.internal.ObjectUtil.checkNotNull;
+
+/**
+ * Combines two {@link DnsQueryLifecycleObserver} into a single {@link DnsQueryLifecycleObserver}.
+ */
+@UnstableApi
+public final class BiDnsQueryLifecycleObserver implements DnsQueryLifecycleObserver {
```

Comment:
You only need to create the list one time when create the factory and the pass the same instance to each observer. So it's not really expensive.... I would prefer to not optimize and introduce just another impl for something that can be done with a generic impl which is more or less kind of free